src/Eval.hs

```diff
@@ -871,43 +865,43 @@ primitiveDefdynamic _ ctx [XObj (Sym (SymPath [] name) _) _ _, value] = do
     Right evaledBody ->
       dynamicOrMacroWith newCtx (\path -> [XObj DefDynamic Nothing Nothing, XObj (Sym path Symbol) Nothing Nothing, evaledBody]) DynamicTy name value
 primitiveDefdynamic _ ctx [notName, _] =
-  pure (evalError ctx ("`defndynamic` expected a name as first argument, but got " ++ pretty notName) (info notName))
+  pure (evalError ctx ("`defndynamic` expected a name as first argument, but got " ++ pretty notName) (xobjInfo notName))
 
 specialCommandSet :: Context -> [XObj] -> IO (Context, Either EvalError XObj)
-specialCommandSet ctx [(XObj (Sym path@(SymPath mod n) _) _ _), value] = do
-  (newCtx, result) <- evalDynamic ctx value
+specialCommandSet ctx [(XObj (Sym path@(SymPath mod n) _) _ _), val] = do
+  (newCtx, result) <- evalDynamic ctx val
   case result of
     Left err -> pure (newCtx, Left err)
     Right evald -> do
       let globalEnv = contextGlobalEnv ctx
       case contextInternalEnv ctx of
         Nothing -> setGlobal newCtx globalEnv evald
         Just env -> setInternal newCtx env evald
-    where setGlobal ctx env value =
+    where setGlobal ctx' env value =
             case lookupInEnv path env of
               Just (_, binder) -> do
-                (ctx', typedVal) <- typeCheckValueAgainstBinder ctx value binder
-                pure $ either (failure ctx) (success ctx') typedVal
+                (ctx'', typedVal) <- typeCheckValueAgainstBinder ctx' value binder
+                pure $ either (failure ctx'') (success ctx'') typedVal
```

Comment:
Seems very reasonable. What's your gut feeling @hellerve ?

Reply:
I think we should use `ctx''` for this one. If it fails here it's because the value and binder didn't type check, but the type information will only be present in `ctx"` since `typeCheckValueAgainstBinder` annotates the types.